scripts.py

```diff
@@ -225,89 +230,369 @@ def run(self, data_source_name, metadata_records, epub_directory, cover_director
-
-        # Get all Works from the Source.
-        works = self._db.query(Work, Identifier, Resource.url)
-        works = works.options(lazyload(Work.license_pools))
-        works = works.join(Work.license_pools).join(LicensePool.data_source).\
-                join(LicensePool.links).join(LicensePool.identifier).\
-                join(Hyperlink.resource).filter(
-                    DataSource.name==source_name,
-                    Hyperlink.rel==Hyperlink.OPEN_ACCESS_DOWNLOAD,
-                    Resource.url.like(u'%.epub')
-                ).all()
+        # Verify the requested DataSources exists.
+        source_names = parser.datasources
+        for name in source_names:
+            source = DataSource.lookup(self._db, name)
+            if not source:
+                raise ValueError('DataSource "%s" could not be found.', name)
+
+        # Get all Works from the DataSources.
+        works_qu = self.base_works_query.filter(
+            DataSource.name.in_(source_names),
+            Hyperlink.rel==Hyperlink.OPEN_ACCESS_DOWNLOAD,
+            Resource.url.like(u'%.epub')
+        )
 
         self.log.info(
-            'Exporting data for %d Works from %s', len(works), source_name
+            'Exporting data for %d Works from DataSources: %s',
+            fast_query_count(works_qu), ','.join(source_names)
         )
 
-        # Transform the works into very basic CSV data for review.
-        rows = list()
-        for work, identifier, url in works:
-            row_data = dict(
-                identifier=identifier.urn.encode('utf-8'),
-                title=work.title.encode('utf-8'),
-                author=work.author.encode('utf-8'),
-                download_url=url.encode('utf-8')
-            )
-            rows.append(row_data)
-
-        # List the works in alphabetical order by title.
-        compare = lambda a,b: cmp(a['title'].lower(), b['title'].lower())
-        rows.sort(cmp=compare)
+        # Transform the works into CSV data for review.
+        rows = list(self.create_row_data(works_qu, parser.source_file))
 
         # Find or create a CSV file in the main app directory.
-        filename = parser.filename
+        filename = parser.output_file
         if not filename.lower().endswith('.csv'):
             filename += '.csv'
         filename = os.path.abspath(filename)
 
-        # Determine whether to append new rows or write over an existing file.
-        open_method = 'w'
-        if parser.append:
-            open_method = 'a'
+        if parser.append and os.path.isfile(filename):
+            existing_rows = None
+            with open(filename) as f:
+                existing_rows = [r for r in csv.DictReader(f)]
+
+            if existing_rows:
+                # Works that are already in the file may have been placed
+                # into different categories in the last run. Prefer newer
+                # categorizations over older ones.
```

Comment:
If I'm reading this right, I think we should prefer older categorizations over newer ones. Otherwise someone who goes in and manually reclassifies a book will have their work overwritten when they update the file..